docs/usecases/testing.md

```diff
@@ -72,3 +72,149 @@ Test aspects are powerful tools for modifying behavior of individual tests or ev
+    equalTo(List(r1, r2, r3, r4, r5, r6, r7, r8, r9))
+  )
+}
+```
+
+When we run out of values in buffer `MockRandom` it falls back to first mode. If we want to we can also clear internal buffers by calling method 
+like `clearInts`.
+
+
+_Mocking Clock_
+
+Sometimes one need to be able to control the flow of time. In most cases you want your unit test to be as fast as possible. Waiting for real time to pass by is a real killer for this. ZIO exposes a `MockClock` in `MockEnvironment` that can control time so we can deterministically and efficiently test effects involving the passage of time.
+
+Example 1
+
+Thanks to call to `MockClock.adjust(1.minute)` we moved the time instantly 1 minute.
+
+```scala
+import zio.test.mock.MockClock
+import zio.duration._
+testM("One can move time very fast") {  
+  for {
+    startTime <- currentTime(TimeUnit.SECONDS)
+    _         <- MockClock.adjust(1.minute)
+    endTime   <- currentTime(TimeUnit.SECONDS)
+  } yield assert(endTime - startTime, isGreaterThanEqualTo(60L))
+}
+```
+
+
+
+Example 2
+
+`MockClock` affects also all code running asynchronously that is scheduled to run after certain time but with caveats to how runtime works.
+
+```scala
+testM("One can control time as he see fit") {
+  for {
+    promise <- Promise.make[Unit, Int]
+    _       <- (ZIO.sleep(10.seconds) *> promise.succeed(1)).fork
+    _       <- MockClock.adjust(10.seconds)
+    readRef <- promise.await
+  } yield assert(1, equalTo(readRef))
+}
+```
+
+The above code creates a write once cell that will be set to "1" after 10 seconds asynchronously from a different thread thanks to call to `fork`. 
+At the end we wait on the promise until it is set. With call to `MockClock.adjust(10.seconds)` we simulate passing of 10 seconds of time.
+Because of it we don't need to wait for the real 10 seconds to pass and thus our unit test can run faster. This is a pattern that will very often be used when `sleep` and `MockClock` are being used for testing of effects that are based on time. The fiber that needs to sleep will be forked and `MockClock` will used to adjust the time so that all expected effects are run in the forked fiber.
+
+*WARNING* 
+Notice that if we don't call `adjust` at all we'll get stuck. `MockClock` doesn't make any progress on its own.
+
+Also it is worth mentioning that adjusting the time on `MockClock` doesn't make us immune to the timing overheads introduced by the runtime and races this introduces. Effects are guaranteed to be waken up not earlier than the argument passed to `sleep` but from there the order of execution since scheduled on different threads is indeterministic and its up to the user code to use tools like `Promise` to guarantee proper sequencing. 
+Below code will be flaky because there is non-zero overhead when switching fibers thus reading of the value might happen before it is set and its change is 
+propagated.
+
+```scala
+testM("THIS TEST WILL FAIL - Sleep and adjust can introduce races") {
```

Comment:
Would be nice to show how to fix this by using a `Promise` that is set _after_ the `sleep` inside the forked fiber; then awaiting on the promise right before `MockClock.adjust`.

Reply:
Isn't it exactly this example?
https://github.com/zio/zio/pull/1800/files#diff-22f248e1a7bb07e86d37efa910d22132R190